File: unity/Rotgrid/Assets/Scripts/Systems/PowerSystem.cs
using UnityEngine;

namespace Rotgrid
{
    /// <summary>
    /// The facility starts dark. Throwing the breaker brings up the lights, the
    /// perk machines, the refit chamber and the traps.
    /// </summary>
    public class PowerSystem : Interactable
    {
        readonly WorldBuilder _world;
        readonly MachineInfo _info;
        public bool On { get; private set; }
        public float ActivatingT;

        public PowerSystem(WorldBuilder world, InteractionSystem interaction, Transform parent)
            : base("power", MapData.PowerSwitchPos + Vector3.up * 1.2f, 2.8f, 2)
        {
            _world = world;
            var go = Props.PowerSwitch(parent);
            go.transform.position = MapData.PowerSwitchPos;
            go.transform.rotation = Quaternion.Euler(0, MapData.PowerSwitchRot, 0);
            _info = go.GetComponent<MachineInfo>();
            world.AddObstacle(MapData.PowerSwitchPos.x, MapData.PowerSwitchPos.z, 1.2f, 0.7f,
                MapData.PowerSwitchPos.y, 2.3f, false);
            interaction.Add(this);
        }

        public override Prompt GetPrompt(GameRun run)
        {
            return On
                ? Prompt.Blocked("MAIN BREAKER", "GRID ONLINE")
                : Prompt.Plain("ACTIVATE POWER", "RESTORES THE FACILITY GRID");
        }

        public override bool Activate(GameRun run)
        {
            if (On) return false;
            On = true;
            ActivatingT = 2.6f;
            _world.SetPowered(true);
            if (_info.pivot != null) _info.pivot.localRotation = Quaternion.Euler(-52f, 0, 0);
            if (_info.glow != null && _info.glow.Length > 0 && _info.glow[0] != null)
                Art.SetEmission(_info.glow[0].material, Util.Hex(0x9fd93a) * 2.2f);
            Game.Audio.Play(Sfx.Power, 1f);
            if (run != null)
            {
                run.Toast("POWER RESTORED", ToastKind.Good, true);
                run.player.AddShake(0.5f);
            }
            return true;
        }

        public void Update(float dt)
        {
            if (ActivatingT > 0f) ActivatingT -= dt;
        }
    }
}

File: src/systems/PowerSystem.js
import { makePowerSwitch, Mats } from '../world/Props.js';
import { Interactable } from './Interaction.js';
import { MAP } from '../world/MapData.js';
import { Audio } from '../core/AudioEngine.js';
import { Emitter } from '../core/Util.js';

/**
 * The facility starts dark. Throwing the breaker brings up perk machines,
 * the refit chamber, traps and the lights.
 */
export class PowerSystem extends Emitter {
  constructor(world, scene, interaction) {
    super();
    this.world = world;
    this.on = false;
    this.activatingT = 0;

    const s = MAP.powerSwitch;
    this.group = makePowerSwitch();
    this.group.position.set(s.x, s.y, s.z);
    this.group.rotation.y = s.rot;
    scene.add(this.group);
    world.addObstacle(s.x, s.z, 1.2, 0.7, s.y, 2.3);

    const self = this;
    this.interactable = interaction.add(new (class extends Interactable {
      prompt() {
        return self.on
          ? { text: 'MAIN BREAKER', sub: 'GRID ONLINE', blocked: true }
          : { text: 'ACTIVATE POWER', sub: 'RESTORES THE FACILITY GRID' };
      }
      activate(ctx) {
        if (self.on) return false;
        self.activate(ctx);
        return true;
      }
    })({ id: 'power', x: s.x, y: s.y, z: s.z, radius: 2.8, priority: 2 }));
  }

  activate(ctx) {
    if (this.on) return;
    this.on = true;
    this.activatingT = 2.6;
    this.world.setPowered(true);
    this.group.userData.lever.rotation.x = -0.9;
    this.group.userData.lamp.material = Mats.emissive(0x9fd93a, 2.2);
    Audio.play('power', { volume: 1 });
    ctx?.toast('POWER RESTORED', 'pts', true);
    ctx?.player.addShake(0.5);
    this.emit('on');
  }

  reset() {
    this.on = false;
    this.world.setPowered(false);
  }

  update(dt) {
    if (this.activatingT > 0) this.activatingT -= dt;
  }
}
